lib/NeuraDialect/Mapping/HeuristicMapping/HeuristicMapping.cpp

```diff
@@ -225,25 +230,51 @@ bool HeuristicMapping::performBacktrack(
 
   // Because we have already thoroughly iterated candidates of no-producer
   // operations. If the current operation is after a no-producer operation, we
-  // need backtrack 2 depths.
+  // need backtrack until we find an operation with a producer.
+  // This is to ensure we do not get stuck in a loop of no-producer operations
+  // without making progress.
+
+  // Example:
+  // <op0, no_producer_op1, no_producer_op2, op3, op4>
+
+  //   |         |                 |        |    |
+
+  //         | loc_0 |         | loc_3 | ...
+  //         | loc_1 |         | loc_4 | ...
+  //         | loc_2 |         | loc_5 | ...
+
+  // 1. If we fail to map op4, we can backtrack to op3 using this
+  // performBacktrack function.
+  // 2. If we fail to map op3, we need to backtrack to op0 to avoid getting
```

Comment:
Why we would have `op0` before `no_producer_op1`? Shouldn't ops be ordered in topological ordering?